From 2127a8c8abdee219ad75ae4f51f4c9d3add2bcdf Mon Sep 17 00:00:00 2001
From: Fabian Vogt <fvogt@suse.de>
Date: Tue, 29 Sep 2026 14:33:48 +0200
Subject: [PATCH] disk/i386/pc/biosdisk: Fix retry of CDROM reads

Make sure dap->blocks is set again before retries, otherwise the retry
may not read the wanted amount of blocks.

This caused boot failures if the kernel file was close enough to the end
of the .iso file: grub_disk_read_small_real attempted to read past the end
of the disk while filling a cache line. The initial read failed, but set
dap->blocks = 0, which made the retry succeed, but not actually return data.

Fixes: https://bugzilla.opensuse.org/show_bug.cgi?id=1282614
Signed-off-by: Fabian Vogt <fvogt@suse.de>
Reviewed-by: Michael Chang <mchang@suse.com>
Reviewed-by: Avnish Chouhan <avnish@linux.ibm.com>
Part-of: <https://gitlab.freedesktop.org/gnu-grub/grub/-/merge_requests/281>
---
 grub-core/disk/i386/pc/biosdisk.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/grub-core/disk/i386/pc/biosdisk.c b/grub-core/disk/i386/pc/biosdisk.c
index 1d6788950..7a3c5b3bb 100644
--- a/grub-core/disk/i386/pc/biosdisk.c
+++ b/grub-core/disk/i386/pc/biosdisk.c
@@ -487,8 +487,16 @@ grub_biosdisk_rw (int cmd, grub_disk_t disk,
 	    return grub_error (GRUB_ERR_WRITE_ERROR, N_("cannot write to CD-ROM"));
 
 	  for (i = 0; i < GRUB_BIOSDISK_CDROM_RETRY_COUNT; i++)
-            if (! grub_biosdisk_rw_int13_extensions (0x42, data->drive, dap))
-	      break;
+	    {
+	      if (! grub_biosdisk_rw_int13_extensions (0x42, data->drive, dap))
+	        break;
+
+	      /*
+	       * The BIOS may have overwritten dap->blocks, e.g. with the number of
+	       * successfully read blocks. Set it again.
+	       */
+	      dap->blocks = size;
+	    }
 
 	  if (i == GRUB_BIOSDISK_CDROM_RETRY_COUNT)
 	    return grub_error (GRUB_ERR_READ_ERROR, N_("failure reading sector 0x%llx "
-- 
2.55.0

